File: summary/ruby/1_selection_sort.rb
def selection_sort(arr)
  n = arr.length

  (0..n - 2).each do |i|
    min_value = arr[i]
    min_index = i

    (i + 1..n - 1).each do |j|
      if min_value > arr[j]
        min_value = arr[j]
        min_index = j
      end
    end

    arr[i], arr[min_index] = arr[min_index], arr[i]
  end

  arr
end

arr = gets.chomp.split.map(&:to_i)
sorted_arr = selection_sort(arr)

puts "#{sorted_arr}"

File: summary/ts/1_selection_sort.ts
function selectionSort(arr: number[]): number[] {
  const n = arr.length;

  for (let i = 0; i < n - 1; i++) {
    // 未ソート部分から最小値を探す
    let minIdx = i;
    for (let j = i + 1; j < n; j++) {
      if (arr[j] < arr[minIdx]) {
        minIdx = j;
      }
    }

    // 未ソート部分の最小値と未ソート部分の先頭を交換
    const temp = arr[i];
    arr[i] = arr[minIdx];
    arr[minIdx] = temp;
  }

  return arr;
}

// 使用例
const unsortedArr = [5, 3, 8, 4, 2];
const sortedArr = selectionSort(unsortedArr);
console.log(sortedArr); // [2, 3, 4, 5, 8]
